apps/src/applab/manageAssets.jsx

```diff
@@ -0,0 +1,138 @@
+var React = require('react');
+var AssetsApi = require('./manageAssets/clientApi');
+var AssetRow = require('./manageAssets/assetRow.jsx');
+
+var errorMessages = {
+  415: 'This type of file is not supported.',
+  500: 'The server responded with an error.'
+};
+
+function getErrorMessage(status) {
+  return errorMessages[status] || 'An unknown error occurred.';
+}
+
+module.exports = React.createClass({
+  propTypes: {
+    assetChosen: React.PropTypes.func,
+    typeFilter: React.PropTypes.string
+  },
+
+  getInitialState: function () {
+    return {
+      assets: null,
+      uploadStatus: ''
+    };
+  },
+
+  componentWillMount: function () {
+    // TODO: Use Dave's client api when it's finished.
+    AssetsApi.ajax('GET', '', function (xhr) {
+      var assets = JSON.parse(xhr.responseText);
+      if (this.props.typeFilter) {
+        assets = assets.filter(function (asset) {
+          return asset.category === this.props.typeFilter;
+        }.bind(this));
+      }
+      this.setState({assets: assets});
+    }.bind(this), function (xhr) {
+      this.setState({uploadStatus: 'Error loading asset list: ' +
+          getErrorMessage(xhr.status)});
+    }.bind(this));
+  },
+
+  fileUploadClicked: function () {
+    var uploader = document.querySelector('#uploader');
+    uploader.click();
+  },
+
+  upload: function () {
+    var file = document.querySelector('#uploader').files[0];
+    var uploadStatus = document.querySelector('#uploadStatus');
+
+    if (file.type && this.props.typeFilter) {
+      var type = file.type.split('/')[0];
+      if (type !== this.props.typeFilter) {
+        this.setState({uploadStatus: 'Only ' + this.props.typeFilter +
+          ' assets can be used here.'});
+        return;
+      }
+    }
+
+    // TODO: Use Dave's client api when it's finished.
+    AssetsApi.ajax('PUT', file.name, function (xhr) {
```

Comment:
It seems like you could be setting the Content-type header on PUT requests, similarly to how it is set in POST requests here https://github.com/code-dot-org/code-dot-org/blob/f0e39bd4393d51664c497c591372024d062803fd/apps/src/applab/appStorage.js#L193 , to eliminate the variability in what the browser will pick.

These are the mime types to use: 
irb(main):001:0> ['jpg', 'jpeg', 'gif', 'png', 'mp3'].map do |type|
irb(main):002:1*   Sinatra::Base.mime_type(type)
irb(main):003:1> end
=> ["image/jpeg", "image/jpeg", "image/gif", "image/png", "audio/mpeg"]

This seems better to me than keeping an ever-growing list of mime types to expect for a given file type as we test in more browsers.